[BE] 이슈 전체 조회 API 구현

kyupid opened the pull request:
안녕하세요. 12팀의 백엔드의 MJ, Kyu 입니다.

### 진행 상황
- 이슈 전체 조회 API 구현

### 고민했던 부분
- 테스트 코드 기반으로 기능을 구현하는데 신경썼습니다.
- 다른 레이어에 의존하지 않는 단위테스트를 구현하는데 집중했습니다.
- 특히 컨트롤러에 대한 단위테스트의 경우 서비스 레이어를 mock으로 구현했기 때문에 컨트롤러 내부의 로직에 대한 단위테스트가 가능합니다.

### 리뷰를 바라는 부분
- 검색을 통해 각각의 레이어에 적합한 테스트 방식을 찾아서 도입했습니다. 혹시 리뷰어님께서 보시기에 더 나은 방식으로 단위테스트가 가능하다면 조언 부탁드립니다.

kyupid committed:
Chore: .gitignore 업데이트

kyupid committed:
[#11] Feat: 깃헙 Oauth로 AccessToken 받아오기

kyupid committed:
[#11] Add: 깃헙 서버에 AccessToken 을 담은 요청

kyupid committed:
[#11] Add: 깃헙에 AccessToken을 가지고 요청하고 응답으로 커스텀 객체로 유저정보를 받음

kyupid committed:
[#11] Feat: Oauth 통해서 JWT 발급

kyupid committed:
[#14] Feat: 모든 api에 대한 Interceptor 적용

kyupid committed:
[#14] Feat: Interceptor를 통해서 JWT 분석 후 특정 API로 데이터를 가져옴

MJbae committed:
프로젝트 세팅 및 로그인 구현

Oauth 통한 JWT 발급 그리고 Interceptor 구현

kyupid committed:
Chore: build.gradle 커밋

kyupid committed:
[#27] Feat: 이슈 전체 조회 API 구현 & 테스트

kyupid committed:
Add: 스키마, 데이터 초기화

MJbae committed:
Merge pull request #29 from kyu-kim-kr/BE/feature

[#27] 이슈 전체 조회 API 구현을 위한 테스트

kyupid committed:
[#27] Add: 예제 데이터 추가

kyupid committed:
[#27] 이슈 전체 조회 API 구현 중 문제

- Entity -> DTO

kyupid committed:
Chore: gradle 관련 파일 업로드

- 프로젝트 빌드 관련

kyupid committed:
Chore: 메인 클래스

Commit:
chore: 프로젝트 설정

kyupid committed:
Add: 유저 정보 받아서 DTO로 넣기

kyupid committed:
Merge branch 'BE/kyu' into BE/feature

kyupid committed:
Add: User 도메인에 기본 생성자 추가

kyupid committed:
Add: Issue 도메인에 List<Label> 추가

kyupid committed:
[#27] Fix: RowMapper 사용 방식 수정

kyupid committed:
[#27] Add: Milestone 도메인, 레포지토리 추가

kyupid committed:
[#27] Add: IssueDto에 MilestoneTitle 가져오기

kyupid committed:
[#27] Add: JsonPropery 어노테이션 추가

kyupid committed:
[#27] Add: Label 도메인, 레포지토리 추가

Commit:
chore: 인터셉터 및 gitignore 설정

Commit:
test: 이슈 전체조회 테스트 구현

Commit:
test: 테스트 목적의 issue 객체에 대해 클래스 분리

Commit:
test: IssueController에 대한 테스트 구현

Commit:
test: test 환경에 InterceptorConfig 추가

kyupid committed:
[#27] Feat: 이슈 전체 조회 API 구현

kyupid committed:
Merge pull request #39 from kyu-kim-kr/BE/feature

[#27] 이슈 전체 조회 API 구현

wheejuni commented on be/src/main/java/com/issuetracker/config/InterceptorConfig.java:
```diff
@@ -0,0 +1,25 @@
+        registry.addInterceptor(new OauthInterceptor(oauthUtil))
+                .addPathPatterns("/api/**")
+                .excludePathPatterns("/api/login/**");
+//                .excludePathPatterns("/api/accommodations/**");
```
삭제합시다~

wheejuni commented on be/src/main/java/com/issuetracker/oauth/LoginController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @GetMapping("/auth")
+    public JwtDto login(@RequestParam String client, @RequestParam String code) {
+        RestTemplate githubRequest = new RestTemplate();
```
이게 왜 컨트롤러에 있죠??

wheejuni commented on be/src/main/java/com/issuetracker/oauth/LoginController.java:
```diff
@@ -0,0 +1,60 @@
+                .header("Accept", "application/json")
+                .body(new GithubAccessTokenRequestDto(
+                        oauthUtil.getClientId(), oauthUtil.getClientSecret(), code, oauthUtil.getRedirectUri()
+                ));
```
이게 왜 컨트롤러에 있죠??(2)

wheejuni commented on be/src/main/java/com/issuetracker/oauth/LoginController.java:
```diff
@@ -0,0 +1,60 @@
+                .header("Authorization", "token " + responseDto.getBody().getAccessToken())
+                .build();
+
+        ResponseEntity<User> user = githubRequest.exchange(request, User.class);
```
GitHub에 요청을 보내고 사용자 정보를 응답받는 로직은 서비스로 이동시켜 주세요.
서비스/컴포넌트 분리를 두려워하지 마세요. 얼마든지 작게 쪼개어도 됩니다. 그리고 그게 더 좋은 설계입니다.

wheejuni commented on be/src/main/java/com/issuetracker/oauth/LoginController.java:
```diff
@@ -0,0 +1,60 @@
+
+        Algorithm algorithm = Algorithm.HMAC256(oauthUtil.getAlgorithmSecret());
+
+        String jwt = JWT.create()
+                .withClaim("id", user.getBody().getId())
+                .withClaim("name", user.getBody().getLogin())
+                .withClaim("avatar_url", user.getBody().getAvatar_url())
+                .withIssuer(oauthUtil.getIssuer())
+                .sign(algorithm);
```
JWT를 생성하는 것도 컨트롤러의 할 일은 아닌 것 같습니다.

wheejuni commented on be/src/main/java/com/issuetracker/oauth/OauthInterceptor.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final JWTVerifier verifier;
+
+    public OauthInterceptor(Oauth oauthUtil) {
```
OAuth는 웹 전반에서 일반명사처럼 사용되는 기술명인데, 이걸 그대로 클래스 이름으로 받은 것은 조금 아쉽네요.

wheejuni commented on be/src/main/java/com/issuetracker/oauth/OauthInterceptor.java:
```diff
@@ -0,0 +1,36 @@
+
+    public OauthInterceptor(Oauth oauthUtil) {
+        Algorithm algorithm = Algorithm.HMAC256(oauthUtil.getAlgorithmSecret());
+        verifier = JWT.require(algorithm)
+                .withIssuer(oauthUtil.getIssuer())
+                .build();
```
제 생각에는 지금 프로젝트 전체에서 한 algorithm을 공유하고 이슈어 이름도 같을 것 같거든요.
그렇다면 알고리즘의 종류나 이슈어 이름 등은 공통 변수로 관리하고 JWTVerifier도 빈으로 등록해서 주입받아 사용하는 것이 좋아 보이는데요, 어떻게 생각하시는지요.

wheejuni commented on be/src/main/java/com/issuetracker/util/RowMappers.java:
```diff
@@ -0,0 +1,52 @@
+import org.springframework.jdbc.core.RowMapper;
+
+public class RowMappers {
+    public static final RowMapper<Issue> ISSUE_ROW_MAPPER = (rs, rowNum) -> {
```
하나하나 `@Bean` 으로 등록하는 건 어떨까요?

wheejuni reviewed (changes requested):
수고 많으셨습니다. 👍 
피드백 확인해주시고요, 전반적으로 `System.out.println()` 을 굉장히 많이 사용하고 있는데 이 부분은 꼭 `Logger` 로 수정해 주시기 바랍니다.
그리고 Java 코딩 컨벤션도 다시 한번쯤 쭉 체크 부탁드려요.
감사합니다.

Commit:
feat: 이슈 생성 구현부 중 labels 제외한 부분 구현

Commit:
feat: 이슈 전체 조회 요청에 labels와 number 부분 추가

Commit:
feat: 담당자와 작성자에 대해 배열 형태로 IssueDto에 전달

Commit:
chore: 불필요한 매개변수 제거

Commit:
feat: 로그인 시 로그인유저가 로그인 테이블에 없다면 자동으로 저장하는 로직 추가

Commit:
feat: 이슈 조건에 따른 검색 기능 구현

Commit:
feat: CORS 처리

Commit:
feat: 이슈 조회 시 closed 속성 추가

Commit:
chore: 초기 데이터 추가
;

Commit:
feat: textColor -> fontLight으로 수정

Commit:
fix: OAuth 적용에 따른 CORS 허용부 수정

Commit:
feat: iOS 로그인 부분 구현

kyupid committed:
[#58] Feat: 이슈 수정 API

kyupid committed:
[#59] Feat: 이슈 삭제, 클로즈여부 API

kyupid committed:
[#60] Feat: 이슈 상세 정보 API

kyupid committed:
[#61] Feat: 오픈/클로즈 이슈 카운트 API

kyupid committed:
[#66] Feat: 마일스톤 전체 조회 API

kyupid committed:
[#66] Feat: 마일스톤 삭제 API

kyupid committed:
[#66] Feat: 마일스톤 수정 API

kyupid committed:
[#66] Feat: 마일스톤 생성 API

MJbae committed:
Merge pull request #62 from kyu-kim-kr/BE/feat-create-issue

[BE] 이슈 생성/검색 API

MJbae committed:
Merge pull request #63 from kyu-kim-kr/BE/feat-edit-issue

[BE] 이슈 수정/삭제/클로즈 여부 API

MJbae committed:
Merge pull request #64 from kyu-kim-kr/BE/feat-detail-issue

[BE] 이슈 상세 정보 API

MJbae committed:
Merge pull request #65 from kyu-kim-kr/BE/feat-count-issue

[BE] 오픈/클로즈 이슈 카운트 API

MJbae committed:
Merge pull request #69 from kyu-kim-kr/BE/feat-milestone

[BE] 마일스톤 전체조회/생성/수정/삭제 API 구현

kyupid committed:
[#66] Feat: 마일스톤 closed 조건에 따른 전체 조회 API

kyupid committed:
Merge pull request #70 from kyu-kim-kr/BE/feat-milestone

Feat: 마일스톤 closed 조건에 따른 전체 조회 API

kyupid committed:
[#67] Feat: 라벨 전체 조회 API

kyupid committed:
[#67] Feat: 라벨 생성 API

kyupid committed:
[#67] Change: label.color_code 의 유니크 속성을 지움

Commit:
Fix: null data 처리 in Issue

kyupid committed:
[#67] Feat: 라벨 수정 API

kyupid committed:
[#67] Feat: 라벨 삭제 API

kyupid committed:
Merge branch 'BE/feature' into BE/feat-label

kyupid committed:
Merge pull request #71 from kyu-kim-kr/BE/feat-label

[BE] 라벨 전체조회/생성/수정/삭제 API 구현

Commit:
fix: UserDto 미생성에 따라 Entity와 Domain의 혼용 발생과 그에  따른 에러 발생

Commit:
feat: 이슈별 코멘트 조회 기능 구현

Commit:
feat: 유저별 코멘트 조회 기능 구현

kyupid committed:
[#85] Feat: 오픈/클로즈 마일스톤 API

kyupid committed:
[#86] Fix: 로그인 DB 중복 데이터 검사 로직

Commit:
feat: 코멘트 생성 구현

Commit:
feat: 코멘트 수정 기능 구현

Commit:
feat: 코멘트 삭제 기능 구현

kyupid committed:
Feat: 담당자/작성자 조회

kyupid committed:
Merge pull request #87 from kyu-kim-kr/kyu/temp

Kyu/temp

kyupid committed:
Merge pull request #88 from kyu-kim-kr/BE/feat-comment

Be/feat comment

kyupid committed:
[#91] Fix/Add: 2021.06.22 회의 후 수정 사항

- [x] 이슈상세조회에 라벨에 font_light 넣기
- [x] 이슈전체조회 assignee에 s하나 붙이기
- [x] 이슈상세조회 assignee 수정
- [x] 이슈전체조회를 했을때  assignees 배열로 반환
- [x] 이슈코멘트카운트 속성 추가

Commit:
Fix: iOS OAuth 로그인 문제 해결

Commit:
fix: 이슈 상세 조회에서 author_id -> author 객체로 수정

Commit:
fix: issue 조회 시, issueRepository에서 엔티티 가져올 때  deleted = false 조건 추가

kyupid committed:
Fix: 이슈 조회 내림차순

kyupid committed:
[#113] Feat: 이미지 업로드 API

kyupid committed:
Merge pull request #114 from kyu-kim-kr/kyu/temp-image

[BE] [#113] Feat: 이미지 업로드 API

kyupid committed:
Merge pull request #122 from kyupid/BE/feature

[BE] 병합

kyupid committed:
Fix: 초기화 데이터 라벨 색상 변경

kyupid committed:
Fix/Add: 테스트 삭제, 필요한 테스트 추가

kyupid committed:
Fix bug: 날짜 매핑하지 못하는 문제 해결

kyupid committed:
Feat: 이모지 상태 전체 조회

kyupid committed:
Merge pull request #129 from kyupid/kyu/feature/get-emojis-status

Feat: 이모지 상태 전체 조회

kyupid committed:
Rename: 변수이름 명확히 하기 위해 수정

kyupid committed:
[#127] Feat: 이모지 달기 구현

kyupid committed:
Merge pull request #133 from kyupid/kyu/feature/emoji

Kyu/feature/emoji 완성

kyupid committed:
[#136] Remove: 필요없는 코드/주석 제거

kyupid committed:
[#136] Fix: 이미지 업로드 리턴타입 수정

kyupid committed:
[FE 요청사항] Fix: 이슈 상제조회시 유저,라벨,마일스톤에 id도 응답하도록 수정

MJbae committed:
Merge pull request #138 from kyupid/kyu/rebase

[BE] 리팩토링용 PR

kyupid committed:
Merge pull request #146 from kyupid/BE/feature

Be/feature
